homeassistant/components/frank_energie/sensor.py

```diff
@@ -0,0 +1,397 @@
+"""Frank Energie current electricity and gas price information service."""
+from __future__ import annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass
+from datetime import timedelta
+from typing import Any
+
+from python_frank_energie.models import Invoices, MarketPrices, MonthSummary
+
+from homeassistant.components.sensor import (
+    SensorDeviceClass,
+    SensorEntity,
+    SensorEntityDescription,
+    SensorStateClass,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    CURRENCY_EURO,
+    EntityCategory,
+    UnitOfEnergy,
+    UnitOfVolume,
+)
+from homeassistant.core import HassJob, HomeAssistant
+from homeassistant.helpers import event
+from homeassistant.helpers.device_registry import DeviceEntryType, DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.typing import StateType
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+from homeassistant.util import dt as dt_util
+
+from .const import (
+    ATTR_TIME,
+    DATA_MONTH_SUMMARY,
+    DOMAIN,
+    SERVICE_NAME_COSTS,
+    SERVICE_NAME_PRICES,
+)
+from .coordinator import FrankEnergieCoordinator
+
+
+@dataclass
+class FrankEnergieEntityDescriptionMixin:
+    """Mixin values for Frank Energie entities."""
+
+    value_fn: Callable[[Invoices | MarketPrices | MonthSummary], StateType]
+
+
+@dataclass
+class FrankEnergieEntityDescription(
+    SensorEntityDescription, FrankEnergieEntityDescriptionMixin
+):
+    """Describes Frank Energie sensor entity."""
+
+    suggested_display_precision: int = 2
+    authenticated: bool = False
+    service_name: str | None = SERVICE_NAME_PRICES
+    attr_fn: Callable[
+        [Invoices | MarketPrices | MonthSummary], dict[str, StateType | list]
+    ] = lambda _: {}
+
+
+SENSOR_TYPES: tuple[FrankEnergieEntityDescription, ...] = (
+    FrankEnergieEntityDescription(
+        key="electricity_markup",
+        translation_key="electricity_markup",
+        native_unit_of_measurement=f"{CURRENCY_EURO}/{UnitOfEnergy.KILO_WATT_HOUR}",
+        value_fn=lambda data: data.electricity.current_hour.total,
+        attr_fn=lambda data: {"prices": data.electricity.asdict("total")},
```

Comment:
I think we should not put json data in state attributes as this isn't user friendly as it's shown in the UI. If a user wants this data, can we maybe use a service call?

Reply:
This is used to generate graphs with apexcharts-card, can we do that using services?

See
https://github.com/bajansen/home-assistant-frank_energie?tab=readme-ov-file#grafiek-voorbeelden

Reply:
The future data is one of the most valuable parts of the integration but I agree this is not the cleanest way of implementing it. I have looked at integrations such as forecast solar for inspiration but was unable to find another method to make this work.